config/tools/config_generator.cpp

```diff
@@ -0,0 +1,935 @@
+                set_param_to_config_file(log_level_key, log_level_val);
+                std::pair<std::string, std::string> key_pair
+                    = create_key_pair(randomize);
+                std::string private_key_key = sentinel_name + "_private_key";
+                std::string private_key_val = key_pair.first;
+                set_param_to_config_file(private_key_key, private_key_val);
+                std::string public_key_key = sentinel_name + "_endpoint";
+                std::string public_key_val = key_pair.second;
+                set_param_to_config_file(public_key_key, public_key_val);
+            }
+        } else if(type == coordinator_count_key) {
+            // Create Coordinators
+            for(size_t i = 0; i < number; i++) {
+                std::string coordinator_name
+                    = "coordinator" + std::to_string(i);
+                std::string endpoint_key = coordinator_name + "_endpoint";
+                std::string endpoint_val = cbdc::network::localhost + ":"
+                                         + std::to_string(get_open_port());
+                set_param_to_config_file(endpoint_key, endpoint_val);
+                std::string raft_endpoint_key
+                    = coordinator_name + "_raft_endpoint";
+                std::string raft_endpoint_val
+                    = cbdc::network::localhost + ":"
+                    + std::to_string(get_open_port());
+                set_param_to_config_file(raft_endpoint_key, raft_endpoint_val);
+                std::string log_level_key = coordinator_name + "_loglevel";
+                std::string log_level_val = _default_debug_level;
+                set_log_level(tmpl_coordinator_log_level, log_level_val);
+                set_param_to_config_file(log_level_key, log_level_val);
+                std::string count_key = coordinator_name + "_count";
+                size_t count_val = number;
+                set_param_to_config_file(count_key, count_val);
+                std::string threads_key = coordinator_name + "_max_threads";
+                size_t threads_val = 1;
+                set_param_to_config_file(threads_key, threads_val);
+            }
+        } else {
+            std::cout << "Warning: Unrecognized component type, " << type
+                      << ", in Two-Phase Commit configuration generation."
+                      << std::endl;
+        }
+    }
+
+    // Method to create all the Atomizer related components for generated
+    // config file
+    void config_generator::create_atomizer_component(const char* type,
+                                                     const size_t number) {
+        size_t randomize_int = 0;
+        find_value(tmpl_randomize_values, randomize_int);
+        bool randomize = randomize_int == 1;
+        auto _default_debug_level = static_cast<std::string>("INFO");
+        find_value(tmpl_default_log_level, _default_debug_level);
+        if(type == shard_count_key) {
```

Comment:
This type of structure is repeated often. Is it possible to refactor this?

My immediate suggestion for how would be to have a method for each component type allowing the body of the for-loop to be a single method call. At that point, it may be possible to refactor a bit further and share the for-loop (dispatching to the routine in the loop body). This also has the added benefit that, should a new type of component be added to the system, ostensibly only the component-specific routine may need to be added to include support.

That will create a few additional helper methods which aren't likely to be called anywhere else, but this is currently pretty hard for me to parse.

Reply:
I understand the first part of the refactor you are suggesting i believe, however, i am not sure what you mean by the second part (how to share the for loop). I will try and think about it more and see if i can understand though. 

NEEDS FOLLOW UP

Reply:
Let me clarify: right now, this code is structured like this:

```c++
if (cond) {
    for(...)
} else if (cond) {
    for(...)
}...
```

The `for`-bodies look like they share a lot of the same structure, which makes it feel like there's potentially a more widely-applicable pattern that would make it easier to maintain this code (by serving as a more direct abstraction). After that simplification, it might make sense to make another reversal:

```c++
for(...) {
    if (cond) {
    } else if (cond) {
    }
}
```

Depending on the type of conditions, this might be equally as or more performant. However, in our case, it's actually likely to be a performance regression (because our conditions are often string-comparison, which isn't likely to have nice low-level performance). But, this isn't the system (where high-performance is a focus), this is just surrounding tooling. All else held equal, specifically in the tooling, I'd prefer to see the code be easier-to-maintain.

Reply:
hmmm. I am still having trouble understanding how the for loop can be shared though since there can be different component numbers for each type of component. You could have a different number of shards or sentinels etc... How can you loop through different numbers for different components in the same for loop?

Reply:
After your refactor, the code looks like this:

```c++
helper1 (count) {
    for (i=0,count)
}

helper2 (count) {
    for (i=0,count)
}

make_components (type, count) {
    if (type == shard) {
        helper1(count)
    } else if ...
}
```

My thought was to have this structure:

```c++
helper1 (index) {
    // routine to generate configs for a single component of a single role
}

helper2 (index) {
    // routine to generate configs for a single component of a single role
}

make_components (type, count) {
    for (i=0,count) {
        if (type == shard) {
            helper1(i)
        } else if ...
    }
}
```

I think that's reasonable (`make_components()` is playing the same role as `create_atomizer_component()` and `create_2pc_component()` in the actual code—which may also be unifiable).

I hope this doesn't feel like busy-work; I think this would be a very helpful tool to have, I'm just hoping to have it be easy for us to maintain.

I'm not entirely convinced that the above refactor is particularly worth it, so think on it a bit before you spend time on implementing it (and send me a PM on Zulip if you want to think through it live or want further clarification).

Reply:
Ok. I think i have implemented what you suggest above. However, in doing so, i came across some confusion i had had and now i remembered. For the 2pc config, this double count for shards and coordinators in the original integration test config still confuses me:

![image](https://user-images.githubusercontent.com/6537956/182133256-be3fddaf-2ba1-4ecd-b076-d1605ec9d98c.png)


I just dont understand why there is an overall count, and then seemingly, a count per shard and coordinator for 2pc. What does the count per shard and coordinator represent? Is it the same count as the general one?

Reply:
Those are raft-replica configs. So, `shard_count` is the total number of logical shards, and `shard0_count` is the number of replicas that will be created in the raft consensus pool backing the first logical shard. Same goes for coordinators.

Reply:
Ok, I am pretty sure i had tested making those numbers different from the ones without a "0" or number in the name and the tests seemed to fail. I will test again. However, that being said, should not the atomizer have such counts too for the atomizer? I do not see such a count in the original atomizer configuration files? There is a total "atomizer_count" but in that case is that the raft replicated number? 

Also, for the 2pc integration tests, making the shard0_count or coordinator0_count does indeed cause the test to cash FYI.

Reply:
Where there are multiple logical shards or coordinators, there can only be one logical atomizer, so the count can only refer to the raft replicas.

Reply:
There is a small issue with the raft replication params i added that i am currently working through FYI.